server/src/main/java/org/elasticsearch/common/util/concurrent/EsExecutors.java

```diff
@@ -319,18 +320,26 @@ private EsExecutors() {}
         public boolean offer(E e) {
             // first try to transfer to a waiting worker thread
             if (tryTransfer(e) == false) {
-                // check if there might be spare capacity in the thread
-                // pool executor
-                int left = executor.getMaximumPoolSize() - executor.getCorePoolSize();
-                if (left > 0) {
+                // optimistically check before queueing whether there is room for more.
+                // have to use `getActiveCount` since those that are active promise to check for tasks before exiting the worker.
+                if (executor.getActiveCount() < executor.getMaximumPoolSize()) {
                     // reject queuing the task to force the thread pool
                     // executor to add a worker if it can; combined
                     // with ForceQueuePolicy, this causes the thread
                     // pool to always scale up to max pool size and we
                     // only queue when there is no spare capacity
                     return false;
                 } else {
-                    return super.offer(e);
+                    boolean offer = super.offer(e);
+                    assert offer;
+                    // check after queuing.
+                    // if all threads are no longer active, see if we can pull item off queue and retry.
```

Comment:
Isn't this 'if at least one thread is no longer active'?

Reply:
Thanks, bad english, corrected in 8399f41d9cd2fddcfddf01d30a3b08c3fa47dd84